### DISCUSSION TITLE
Airflow 2 upgrade introduces Jinja errors in DAG runs, but templates aren't used

### DISCUSSION THREAD

**Discussion by rschoenbeck** [apache/airflow#36600] [Q&A] [answered]
We're being forced to upgrade from Airflow 1.10.15 to 2.6.3 due to Google Cloud Composer dropping support for Airflow < 2.0, but are encountering Jinja errors on previously-functioning DAGs. 

The DAGs do not actually use Jinja templating.

Here's an example error from a run of a previously-functioning diagnostics DAG. It's unclear what the field `value` refers to -- again, our code uses no Jinja templates, and the template here looks like a logging template of some kind used by another process.

```
[2024-01-04, 19:10:43 UTC] {abstractoperator.py:594} ERROR - Exception rendering Jinja template for task 'diagnostics-v1', field 'value'. Template: '[%(asctime)s] {%(filename)s:%(lineno)d} %(levelname)s - %(message)s'
Traceback (most recent call last):
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/models/abstractoperator.py", line 586, in _do_render_template_fields
    rendered_content = self.render_template(
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/template/templater.py", line 155, in render_template
    template = jinja_env.from_string(value)
  File "/opt/python3.8/lib/python3.8/site-packages/jinja2/environment.py", line 1105, in from_string
    return cls.from_code(self, self.compile(source), gs, None)
  File "/opt/python3.8/lib/python3.8/site-packages/jinja2/environment.py", line 768, in compile
    self.handle_exception(source=source_hint)
  File "/opt/python3.8/lib/python3.8/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<unknown>", line 1, in template
jinja2.exceptions.TemplateSyntaxError: tag name expected
```

And the DAG is below.

`kivaflow` is an internal library that just essentially just generates a KubernetesPodOperator with some default settings (and also does not use Jinja templating).

```from airflow import DAG

from kivaflow import dags
from kivaflow.affinity import Nodes
from kivaflow.image import Images
from kivaflow.pods import PodData, get_pod

###############################################################################
# Configuration
###############################################################################

dag_name = "diagnostics_v1_0"
tag = "DSCI-1674"

delay = 30  # minutes, after the hour
window = 1  # hours

dag = DAG(
    dag_name,
    default_args=dags.get_args(),
    schedule_interval=f"{delay} */{window} * * *",
    catchup=False
)

###############################################################################
# DAG setup
###############################################################################

diagnostic_entrypoint = f"""
from diagnostics import driver
driver.run_diagnostics() 
"""

monitor_data = PodData(
    "dags",
    dag,
    Images.base,
    tag,
    diagnostic_entrypoint,
    "diagnostics-v1",
    Nodes.standard_mem,
    "4G",
    1.0,
)
monitor = get_pod(monitor_data)

###############################################################################
# DAG steps
###############################################################################

monitor
```

Anyone seen this before? Any help appreciated.

**Comment by potiuk** [comment]
Yes. Likely your log template is wrong/old. There are a number of log templates in config https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#colored-console-log - likely your airflow.cfg or airflow variables contain some other format. The error you see is likely not the root cause - it's likely somethign trying to write to log and encountering tag that the Airflow 2. logger does not understand.


Unfortunately you are very late with the migration - 2.5 years after Airflow 1 reached EOL so you won't get likely much help here and the on

I recommend you to install completely new, fresh instance and add your apply any configuration change syou had in Airflow 1.10 after carefully looking at what was in the original code and comparing your configuration you had and getting to understand why you applied it in the first place. 

Unfortunately, after waiting more than 3 years, there is almost no-one who remember anything about Airflow 1.10. So you are mostly on your own to solve those kind of problems.  You can also refer to this document that was written more than 3 years ago https://airflow.apache.org/docs/apache-airflow/stable/howto/upgrading-from-1-10/index.html - if there is any place where "migration" knowledge is kept - that's the place.

I guess that's the price to pay for not upgrading for such a long time, I strongly recommend to follow different patterns in the future. Airflow 2 follows semver and genearally you are expected to upgrade regularly and it should be much easier that 1 -> 2 migration. By doing it you make sure you update to latest security fixes as well. Just a comment that in ~ 2 years after current EU and US regulations about security will be put in place will be pretty mandatory for everyone unless your company and their management will want to be prone to penalties - both financial and in some cases criminal, so my recommendation is start doing it now to get used to this process.

**Comment by rschoenbeck** [comment]
Thanks for the reply. To be clear, we're working with the default Airflow configuration that comes with a Cloud Composer 2.5.3 image. So it is unlikely that the issue is coming from the config, unless Google messed it up.

We would have upgraded sooner, but we're a very small team, the 2.0 upgrade required a decent amount of work for our deployment, and Google continued to offer support for older Airflow images through releases of Cloud Composer, so we figured we were OK.

I was able to create a very simple test DAG (below), and that seemed to run OK. So my guess is that probably the templater in Airflow 2.6.3 doesn't like something in the formatting of the kubernetes pod operator arguments that Airflow 1.10.15 was OK with...I've already tried the most obvious suspects, so next step is probably breaking it down piece by piece and trying it repeatedly.

```
dag_name = "test_dag"
tag = "DSCI-1674"

delay = 30  # minutes, after the hour
window = 1  # hours

default_args = {
    "owner": "airflow",
    "depends_on_past": False,
    "catchup": False,
    "description": "it's a test DAG",
    "start_date": datetime(2020, 1, 1),
    "email": ["xxx-yyy"],
    "email_on_failure": True,
    "email_on_retry": False,
    "retries": 3,
    "max_active_runs": 1,
    "retry_delay": timedelta(minutes=5),
}

dag = DAG(
    dag_name,
    default_args=default_args,
    schedule_interval=f"{delay} */{window} * * *",
    catchup=False,
    tags=["test"]
)

t1 = BashOperator(
        task_id='print_date',
        bash_command='date',
        dag=dag
)

t1
```

**Reply by potiuk** [reply]
Well. KPO has changed A LOT since 1.10. We are at 7.* version of Kubernetes Provider so there were 7 versions with breaking changes since the beginning - Airflow 2 splits into "core" airflow versioned separately and "providers" versioned separately are versioned separatly, can individually introduce breaking changes (but you can also downgrate the providers to previous versions if you want to keep providers compatible, and upgrade airflow).

So. yes for like 1000% your generator will not work with current KPO.

You can take a look at Changelog of the KPO provoder (going back to 1.0) https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/index.html if you want to see details on how it changed.

**Comment by rschoenbeck** [comment, marked as answer]
Just to close the loop here: it ended up being an environment variable change in the Cloud Composer 2 environment that made Jinja wig out, due to the variable's value having percentage signs -- so, not even an Airflow thing. We didn't actually use said variable anymore, so dropping it fixed the problem. 

(As suggested above, the Kubernetes changes are the bigger issue in a Cloud Composer 2 migration, but that's a separate thing entirely.)

Thanks again for the advice. Hope this helps someone else in the same migration boat, but probably not too likely to happen. :)
